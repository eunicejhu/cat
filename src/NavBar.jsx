import React from "react";
import { Link } from "react-router-dom";

const NavBar = () => (
  <nav>
    <li>
      <Link to="/">Home</Link>
    </li>
  </nav>
);

export default NavBar;
