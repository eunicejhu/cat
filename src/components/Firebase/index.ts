import app from "firebase/app";
import "firebase/auth";
import FirebaseContext from "./context";
import useFirebase from "./useFirebase";

//TODO: .env.production, .env.development
const config = {
    apiKey: process.env.REACT_APP_API_KEY,
    authDomain: process.env.REACT_APP_AUTH_DOMAIN,
    databaseURL: process.env.REACT_APP_DATABASE_URL,
    projectId: process.env.REACT_APP_PROJECT_ID,
    storageBucket: process.env.REACT_APP_STORAGE_BUCKET,
    messagingSenderId: process.env.REACT_APP_MESSAGING_SENDER_ID,
    appId: process.env.REACT_APP_APP_ID,
    measurementId: process.env.REACT_APP_MEASUREMENT_ID,
};

class Firebase {
    auth: app.auth.Auth;
    constructor() {
        app.initializeApp(config);
        this.auth = app.auth();
        this.auth.useEmulator("http://localhost:9099/");
    }

    /**
     * Auth API
     */
    doCreateUserWithEmailAndPassword = (email: string, password: string) =>
        this.auth.createUserWithEmailAndPassword(email, password);
    doSignInWithEmailAndPassword = (email: string, password: string) =>
        this.auth.signInWithEmailAndPassword(email, password);
    doSignOut = () => this.auth.signOut();
    doPasswordReset = (email: string) =>
        this.auth.sendPasswordResetEmail(email);
    doPasswordUpdate = (newPassword: string) =>
        this.auth.currentUser?.updatePassword(newPassword);
}

export default Firebase;
export { FirebaseContext, useFirebase };
