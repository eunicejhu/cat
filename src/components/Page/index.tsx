import styled from "styled-components";
export const StyledForm = styled.form`
    display: flex;
    flex-direction: column;
    * {
        margin: var(--margin) 0;
        border-radius: var(--borderRadius);
    }
`;

export const StyledPage = styled.div`
    text-align: center;
    width: 30%;
    min-width: 300px;
    margin: auto;
`;

export const StyledSubmitButton = styled.button`
    width: 50%;
    min-width: 150px;
    margin: auto;
    padding: 10px;
`;
