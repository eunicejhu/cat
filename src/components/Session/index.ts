import AuthUserContext from "./context";
import useAuthUser from "./useAuthUser";
import useAuthorization from "./useAuthorization";

export { AuthUserContext, useAuthUser, useAuthorization };
