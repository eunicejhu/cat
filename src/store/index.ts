import { configureStore } from "@reduxjs/toolkit";
import postsReducer, { Post } from "../features/posts/postsSlice";
import usersReducer from "../features/users/usersSlice";
import notificationsReducer, {
  Notification,
} from "../features/notifications/notificationsSlice";
import { useDispatch } from "react-redux";

const store = configureStore({
  reducer: {
    posts: postsReducer,
    users: usersReducer,
    notifications: notificationsReducer,
  },
});

export type AppDispatch = typeof store.dispatch;
export const useAppDispatch = () => useDispatch<AppDispatch>();

export type State = ReturnType<typeof store.getState>;
// selectors
export const selectAllPosts = (state: State): Post[] => state.posts.data;
export const selectPostById = (
  state: State,
  postId: string
): Post | undefined => state.posts.data.find((post) => post.id === postId);
export const selectAllNotifications = (state: State): Notification[] =>
  state.notifications;

export default store;
