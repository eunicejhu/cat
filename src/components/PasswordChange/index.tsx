import React, { useState } from "react";
import Firebase, { useFirebase } from "../Firebase";
import Input from "../input/Input";
import { StyledForm, StyledSubmitButton } from "../Page";
const PasswordChangeForm = () => {
    const [password1, setPassword1] = useState("");
    const [password2, setPassword2] = useState("");
    const [error, setError] = useState("");
    const [success, setSuccess] = useState(false);
    const firebase = useFirebase() as Firebase;
    const onChangeInput = (e: React.SyntheticEvent) => {
        const { value, name } = e.target as typeof e.target & {
            value: string;
            name: string;
        };
        if (name === "password1") {
            setPassword1(value);
        } else {
            setPassword2(value);
        }
    };
    const onSubmit = async (e: React.SyntheticEvent) => {
        e.preventDefault();
        try {
            await firebase.doPasswordUpdate(password1);
            setPassword1("");
            setPassword2("");
            setError("");
            setSuccess(true);
            setTimeout(() => setSuccess(false), 1000);
        } catch (error) {
            setError(error.message);
        }
    };
    const canSubmit =
        Boolean(password1) && Boolean(password2) && password1 === password2;
    return (
        <StyledForm>
            <h2>Reset Password</h2>
            <Input
                type="password"
                name="password1"
                placeholder="Password"
                data-testid="password1"
                value={password1}
                onChange={onChangeInput}
            ></Input>
            <Input
                type="password"
                name="password2"
                placeholder="Confirm Password"
                data-testid="password2"
                value={password2}
                onChange={onChangeInput}
            ></Input>
            <p>{error}</p>
            <StyledSubmitButton
                type="submit"
                data-testid="submit"
                disabled={!canSubmit}
                onClick={onSubmit}
            >
                {success ? "Success!" : "Reset Password"}
            </StyledSubmitButton>
        </StyledForm>
    );
};

export default PasswordChangeForm;
