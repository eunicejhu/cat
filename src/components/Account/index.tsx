import React from "react";
import PasswordChangeForm from "../PasswordChange";
import { PasswordForgetForm } from "../PasswordForget";
import { StyledPage } from "../Page";
import { useAuthorization } from "../Session";
const Account = () => {
    const authUser = useAuthorization();
    return authUser ? (
        <StyledPage>
            <h2>Reset By Email</h2>
            <PasswordForgetForm />
            <PasswordChangeForm />
        </StyledPage>
    ) : null;
};

export default Account;
