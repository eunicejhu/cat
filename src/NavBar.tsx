import React from "react";
import { Link } from "react-router-dom";
import styled from "styled-components";
import SignOutButton from "./components/SignOut";

import { useTheme, Themes, Mode } from "./components/Theme";

import {
    LANDING,
    NOTIFICATIONS,
    SIGN_IN,
    HOME,
    ACCOUNT,
    ADMIN,
} from "./constants/routes";

interface StyledLinkProps {
    themes: Themes;
    mode: Mode;
}
const StyledLink = styled(Link)<StyledLinkProps>`
    ${({ themes, mode }) => {
        return ` --btn-text-hover:  ${themes[mode].btnTextHover};
            --btn-bg-hover: ${themes[mode].btnBackgroundHover};
            --color-bg: ${themes[mode].background} ;
            --color-text:${themes[mode].text};
            --spacing:${themes[mode].spacing};
            `;
    }}
    && {
        text-decoration: none;
        color: var(--color-text);
        background-color: var(--color-bg);
        border: none;
        padding: var(--spacing);
    }

    &&:hover {
        color: var(--btn-text-hover);
        cursor: pointer;
        background-color: var(--btn-bg-hover);
    }
`;

const LinkWithTheme = ({ to, text }: { to: string; text: string }) => {
    const { themes, mode } = useTheme();
    return (
        <li>
            <StyledLink themes={themes} mode={mode} to={to}>
                {text}
            </StyledLink>
        </li>
    );
};

const NavBar = () => {
    return (
        <nav>
            <LinkWithTheme to={LANDING} text="Landing" />
            <LinkWithTheme to={SIGN_IN} text="Sign In" />
            <LinkWithTheme to={HOME} text="Home" />
            <LinkWithTheme to={ACCOUNT} text="Account" />
            <LinkWithTheme to={ADMIN} text="Admin" />
            <LinkWithTheme to={NOTIFICATIONS} text="Notifications" />
            <li>
                <SignOutButton />
            </li>
        </nav>
    );
};

export default NavBar;
