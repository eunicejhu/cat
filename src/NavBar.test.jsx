import React from "react";
import { render, fireEvent } from "@testing-library/react";
import NavBar from "./NavBar";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import store from "./store/index";
import { AuthUserContext } from "./components/Session";

const withBrowserRouterAndStoreAndAuth = (
    ui,
    authUser = null,
    route = "/initial"
) => {
    window.history.pushState({}, "", route);
    return (
        <AuthUserContext.Provider value={authUser}>
            <BrowserRouter>
                <Provider store={store}>{ui}</Provider>
            </BrowserRouter>
        </AuthUserContext.Provider>
    );
};
const mockAuthUser = {};
describe("NavBarNonAuth", () => {
    it("render correctly", () => {
        const ui = withBrowserRouterAndStoreAndAuth(<NavBar />);
        const { getByText } = render(ui);
        expect(getByText(/Landing/g)).toBeInTheDocument();
        expect(getByText(/Sign In/g)).toBeInTheDocument();
    });
});

describe("NavBarAuth", () => {
    it("renders correctly", () => {
        const ui = withBrowserRouterAndStoreAndAuth(<NavBar />, mockAuthUser);
        const { asFragment } = render(ui);
        expect(asFragment()).toMatchSnapshot();
    });

    it("click Home, location pathname set to /home", () => {
        const ui = withBrowserRouterAndStoreAndAuth(<NavBar />, mockAuthUser);
        const { getByText } = render(ui);
        fireEvent.click(getByText(/Home/i));
        expect(window.location.pathname).toBe("/home");
    });

    it("click Notifications, location pathname set to /notifications", () => {
        const ui = withBrowserRouterAndStoreAndAuth(<NavBar />, mockAuthUser);
        const { getByText } = render(ui);
        fireEvent.click(getByText(/Notifications/i));
        expect(window.location.pathname).toBe("/notifications");
    });
});
