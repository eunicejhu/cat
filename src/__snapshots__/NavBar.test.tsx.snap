// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<nav>
  <li>
    <a
      href="/"
      onClick={[Function]}
    >
      Home
    </a>
  </li>
  <li>
    <a
      href="/cats"
      onClick={[Function]}
    >
      Cats
    </a>
  </li>
</nav>
`;
